src/main/java/slimeknights/tconstruct/library/recipe/alloy/recipe/AlloyRecipe.java

```diff
@@ -0,0 +1,147 @@
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import slimeknights.mantle.recipe.FluidIngredient;
+import slimeknights.mantle.recipe.ICommonRecipe;
+import slimeknights.mantle.recipe.ICustomOutputRecipe;
+import slimeknights.mantle.recipe.RecipeHelper;
+import slimeknights.tconstruct.TConstruct;
+import slimeknights.tconstruct.library.recipe.RecipeTypes;
+import slimeknights.tconstruct.smeltery.TinkerSmeltery;
+import slimeknights.tconstruct.library.recipe.alloy.inventory.IAlloyInventory;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+@AllArgsConstructor
+public class AlloyRecipe implements IAlloyRecipe {
+  @Getter
+  protected final ResourceLocation id;
+  @Getter
+  protected final String group;
+  @Getter
+  protected final NonNullList<FluidIngredient> fluidIngredients;
+  @Getter
+  private final FluidStack output;
+  @Getter
+  private final int temperature;
+
+  @Override
+  public boolean matches(IAlloyInventory inv, World worldIn) {
+    boolean match = false;
+    for (FluidIngredient fluidIngredient : fluidIngredients) {
+      for (FluidStack candidate : inv.getFluidStacks()) {
+        match &= fluidIngredient.test(candidate);
+      }
+    }
+    match &= inv.getFluid() == output.getFluid();
+    return match;
+  }
+
+  @Override
+  public int getTemperature(IAlloyInventory inv) {
+    return temperature;
+  }
+
+  @Override
+  public FluidStack getOutput(IAlloyInventory inv) {
+    return output.copy();
+  }
+
+  @Override
+  public IRecipeSerializer<?> getSerializer() {
+    return TinkerSmeltery.alloySerializer.get();
+  }
+
+  /**
+   * Gets a list of valid fluid inputs for this recipe, for display in jei;
+   * TODO: fix
+   */
+  public List<FluidStack> getFluids() {
+    List<FluidStack> fluids = new ArrayList<>();
+    for (FluidIngredient fluidIngredient : fluidIngredients) {
+      fluids.addAll(fluidIngredient.getFluids());
+    }
+    return fluids;
+  }
+
+  /**
+   * Serializer for {@link AlloyRecipe}
+   */
+  public static class Serializer extends ForgeRegistryEntry<IRecipeSerializer<?>>
+    implements IRecipeSerializer<AlloyRecipe> {
+
+    @Override
+    public AlloyRecipe read(ResourceLocation recipeId, JsonObject json) {
+      String group = JSONUtils.getString(json, "group", "");
+      NonNullList<FluidIngredient> fluidIngredients = readFluidIngredients(JSONUtils.getJsonArray(json, "ingredients"));
+      FluidStack output = RecipeHelper.deserializeFluidStack(JSONUtils.getJsonObject(json, "result"));
+      int temperature = JSONUtils.getInt(json, "temperature");
+      return new AlloyRecipe(recipeId, group, fluidIngredients, output, temperature);
+    }
+
+    private static NonNullList<FluidIngredient> readFluidIngredients(JsonArray array) {
+      NonNullList<FluidIngredient> nonNullList = NonNullList.create();
```

Comment:
Look into `JsonHelper::readList`.

Reply:
Assuming you mean `slimeknights.mantle.util.JsonHelper` there is no `readList`
function, unless you mean `parseList`?

Reply:
Sure, same thing